### PR TITLE
executor: check error message and refine LOAD DATA logic

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #40499

Problem Summary:

### What is changed and how it works?

- Add specific error types for LOAD DATA, and display the error to user (TODO: check data error is also displayed)
- refine the producer/comsumer pattern, use errgroup instead of multiple channels with magic size
- remove some public member and method, rename and reorder functions

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/loadremotetest/error_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	s.tk.MustExec("CREATE TABLE t (i INT, s varchar(32));")
+
+	err = s.tk.ExecToErr("LOAD DATA INFILE 'gs://1' INTO TABLE t (wrong)")
+	checkClientErrorMessage(s.T(), err, "ERROR 1054 (42S22): Unknown column 'wrong' in 'field list'")
+	// This behaviour is different from MySQL
+	err = s.tk.ExecToErr("LOAD DATA INFILE 'gs://1' INTO TABLE t (i,i)")
+	checkClientErrorMessage(s.T(), err, "ERROR 1110 (42000): Column 'i' specified twice")
+	err = s.tk.ExecToErr("LOAD DATA INFILE 'gs://1' INTO TABLE t (@v) SET wrong=@v")
+	checkClientErrorMessage(s.T(), err, "ERROR 1054 (42S22): Unknown column 'wrong' in 'field list'")
+
```
Comment (fzzf678) [inline executor/loadremotetest/error_test.go:52]:
```suggestion
```

File: executor/load_data.go [hunk 2]
```diff
@@ -390,104 +348,188 @@ func (e *LoadDataInfo) initLoadColumns(columnNames []string) error {
...
+		}
+
+	TrySendTask:
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-checkKilled.C:
+			if atomic.CompareAndSwapUint32(&e.Ctx.GetSessionVars().Killed, 1, 0) {
+				logutil.Logger(ctx).Info("load data query interrupted quit data processing")
+				return ErrQueryInterrupted
```
Comment (D3Hunter) [inline executor/load_data.go:480]: `close(e.commitTaskQueue)` before return?

File: executor/load_data.go [hunk 3]
```diff
@@ -514,74 +556,52 @@ func (e *LoadDataInfo) CommitOneTask(ctx context.Context, task commitTask) error
 	return err
 }
 
-// commitWork commit batch sequentially
-func (e *LoadDataInfo) commitWork(ctx context.Context) error {
+// CheckAndInsertOneBatch is used to commit one transaction batch full filled data
+func (e *LoadDataWorker) CheckAndInsertOneBatch(ctx context.Context, rows [][]types.Datum, cnt uint64) error {
+	if e.stats != nil && e.stats.BasicRuntimeStats != nil {
+		// Since this method will not call by executor Next,
+		// so we need record the basic executor runtime stats by ourself.
```
Comment (D3Hunter) [inline executor/load_data.go:563]:
```suggestion
		// so we need record the basic executor runtime stats by ourselves.
```

File: executor/load_data.go [hunk 4]
```diff
@@ -606,7 +626,10 @@ func (e *LoadDataInfo) ReadRows(ctx context.Context, parser mydump.Parser) error
 			if errors.Cause(err) == io.EOF {
 				return nil
 			}
-			return err
+			return ErrLoadDataCantRead.GenWithStackByArgs(
+				err.Error(),
+				"Only the following formats delimited text file (csv, tsv), parquet, sql are supported. Please provide the valid source file(s)",
```
Comment (D3Hunter) [inline executor/load_data.go:631]: should we include help message in [doc](https://docs.pingcap.com/tidb/dev/error-codes), not in error message? seems most tidb error don't have msg for how to handle it.

File: executor/load_data.go [hunk 5]
```diff
@@ -43,12 +44,13 @@ import (
...
 	"github.com/pingcap/tidb/util/logutil"
 	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
 )
 
 const (
 	// LoadDataFormatSQLDump represents the data source file of LOAD DATA is
 	// mydumper-format DML file
-	LoadDataFormatSQLDump = "sqldumpfile"
+	LoadDataFormatSQLDump = "sql file"
```
Comment (D3Hunter) [inline executor/load_data.go:53]: use `sql`?

File: executor/load_data.go [hunk 6]
```diff
@@ -132,36 +147,36 @@ func (e *LoadDataExec) loadFromRemote(
 	}
 	s, err := storage.New(ctx, b, opt)
 	if err != nil {
-		return err
+		return ErrLoadDataCantAccess
```
[the remark this answers is not in the document]
Comment (lance6716) [reply executor/load_data.go:150]: I'll discuss with PM about it later

File: executor/load_data.go [hunk 7]
```diff
@@ -132,36 +147,36 @@ func (e *LoadDataExec) loadFromRemote(
...
 	}
 	fileReader, err := s.Open(ctx, filename)
 	if err != nil {
-		return err
+		return ErrLoadDataCantRead.GenWithStackByArgs(getMsgFromBRError(err), "Please check the INFILE path is correct")
```
[the remark this answers is not in the document]
Comment (lance6716) [reply executor/load_data.go:154]: I'll discuss with PM about it later

File: executor/load_data.go [hunk 8]
```diff
@@ -390,104 +348,188 @@ func (e *LoadDataInfo) initLoadColumns(columnNames []string) error {
...
+	for idx, colName := range columnNames {
+		mapping[strings.ToLower(colName)] = idx
 	}
 
-	return columns
-}
+	for _, col := range cols {
+		idx := mapping[col.Name.L]
+		reorderedColumns[idx] = col
+	}
```
[the remark this answers is not in the document]
Comment (lance6716) [reply executor/load_data.go:374]: I just copy the old code, will check this comment in next commit

File: executor/load_data.go [hunk 9]
```diff
@@ -390,104 +348,188 @@ func (e *LoadDataInfo) initLoadColumns(columnNames []string) error {
...
+	TrySendTask:
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-checkKilled.C:
+			if atomic.CompareAndSwapUint32(&e.Ctx.GetSessionVars().Killed, 1, 0) {
+				logutil.Logger(ctx).Info("load data query interrupted quit data processing")
+				return ErrQueryInterrupted
+			}
+			goto TrySendTask
```
[the remark this answers is not in the document]
Comment (lance6716) [reply executor/load_data.go:482]: I prefer this goto, it's very easy to understand 😄

File: executor/load_data.go (hunk 4 above)
Comment (lance6716) [reply executor/load_data.go:631]: I'll discuss with PM about it later

File: executor/load_data.go (hunk 5 above)
Comment (lance6716) [reply executor/load_data.go:53]: I'll discuss with PM about it later

**Review by D3Hunter — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**